standard/tabs.lua

```diff
@@ -0,0 +1,232 @@
+end
+
+---@param args table
+---@param options {allowThis2: boolean?, removeEmptyTabs: boolean?}
+---@return {name: string?, link: string?, content: string|Html?, tabs: string|Html?, this: boolean}[]
+function Tabs._readArguments(args, options)
+	local tabArgs = {}
+	local tabIndex = 1
+	local this = tonumber(args.This)
+	local this2 = tonumber(args.This2)
+
+	while args['name' .. tabIndex] or args['link' .. tabIndex] do
+		if args['content' .. tabIndex] or not options.removeEmptyTabs then
+			table.insert(tabArgs, {
+				name = Table.extract(args, 'name' .. tabIndex),
+				link = Table.extract(args, 'link' .. tabIndex),
+				content = Table.extract(args, 'content' .. tabIndex),
+				tabs = Table.extract(args, 'tabs' .. tabIndex),
+				this = this == tabIndex or (options.allowThis2 and this2 == tabIndex),
+			})
+		end
+		tabIndex = tabIndex + 1
+	end
+
+	if Logic.readBool(args.returnIfEmpty) then
+		return tabArgs
+	end
+
+	assert(Logic.isNotEmpty(tabArgs), 'You are trying to add a "Tabs" template without arguments for names nor links')
+
+	return tabArgs
+end
+
+---@param tabArgs {name: string?, link: string?, content: string|Html?, tabs: string|Html?, this: boolean}[]
+function Tabs._setThis(tabArgs)
+	if Array.any(tabArgs, Operator.property('this')) then return end
+
+	local fullPageName = mw.title.getCurrentTitle().prefixedText
+	local this
+
+	-- Finds the link that is a prefix of the current page. If there are more than one, choose the longest, then first.
+	-- For example, if the current page is ab/cd/e3, then among
+	--   ab/cd/e1
+	--   ab/cd/e2
+	--   ab/cd/e
+	--   ab/cd
+	--   ab/cg
+	--   ab
+	-- it will pick ab/cd.
+	local maxLinkLength = -1
+
+	Array.forEach(tabArgs, function (tab, tabIndex)
+		local link = tab.link
+		if not link then return end
+		link = link:gsub('_', ' ')
+		local linkLength = string.len(link)
+		local charAfter = string.sub(fullPageName, linkLength + 1, linkLength + 1)
+		local pagePartial = string.sub(fullPageName, 1, linkLength)
+		if pagePartial == link and (charAfter == '/' or charAfter == '') and linkLength > maxLinkLength then
+			maxLinkLength = linkLength
+			this = tabIndex
+		end
+	end)
+
+	if not this then return end
+
+	tabArgs[this].this = true
+end
+
+---@param hasContent boolean
+---@param hybridTabs boolean
+---@param noPadding boolean
+---@return Html|string
+function Tabs._buildContentDiv(hasContent, hybridTabs, noPadding)
+	if hasContent then
+		local contentDiv = mw.html.create('div')
+			:addClass('tabs-content')
+		if hybridTabs then
+			contentDiv
+				:css('border-style', 'none !important')
+				:css('padding', '0 !important')
+		elseif noPadding then
+			contentDiv
+				:css('padding', '0 !important')
+		end
+		return contentDiv
+	end
+
+	local style = ''
+	if hybridTabs then
+		style = 'border-style:none !important; padding:0 !important;'
+	elseif noPadding then
+		style = 'padding:0 !important;'
+	end
+	return '\n<div class="tabs-content" style="' .. style .. '">'
```

Comment:
Is this leftover?

Reply:
no it is intended
tabs dynamic has 2 valid input patterns
one which uses purely the module and one where you have the contents outside of it (using `Template:Tabs dynamic/tab`)
for that the outer div and content div have to be returned open and hence i can not really use the mw.html stuff
example usage: https://liquipedia.net/starcraft2/index.php?title=IEM_Katowice/2024&action=edit&section=5

Reply:
Ah right, that's a bit annoying.